kotlinx-coroutines-core/jvm/src/internal/ExceptionsConstructor.kt

```diff
@@ -32,13 +32,29 @@ internal fun <E : Throwable> tryCopyException(exception: E): E? {
 
 private fun <E : Throwable> createConstructor(clz: Class<E>): Ctor {
     val nullResult: Ctor = { null } // Pre-cache class
-    // Skip reflective copy if an exception has additional fields (that are usually populated in user-defined constructors)
+    // Skip reflective copy if an exception has additional fields (that are typically populated in user-defined constructors)
     if (throwableFields != clz.fieldsCountOrDefault(0)) return nullResult
     /*
-    * Try to reflectively find constructor(), constructor(message, cause), constructor(cause) or constructor(message).
-    * Exceptions are shared among coroutines, so we should copy exception before recovering current stacktrace.
-    */
-    val constructors = clz.constructors.sortedByDescending { it.parameterTypes.size }
+     * Try to reflectively find constructor(message, cause), constructor(message), constructor(cause), or constructor()
+     * Exceptions are shared among coroutines, so we should copy exception before recovering current stacktrace.
+     *
+     * By default, Java's reflection iterates over ctors in the source-code order and the sorting is stable,
+     * so in comparator we are picking the message ctor over the cause one to break such implcit dependency on the source code.
+     */
+    val constructors = clz.constructors.sortedByDescending {
+        // (m, c) -> 3
+        // (m) -> 2
+        // (c) -> 1
+        // () -> 0
+        val params = it.parameterTypes // cloned array
+        when (params.size) {
+            2 -> 3
+            1 -> if (params[0] == String::class.java) 2 else 1
+            0 -> 0
+            else -> -1
+        }
```

Comment:
I agree with the change, but I'd like to propose a stylistic edit: https://gist.github.com/dkhalanskyjb/2c2b4362920a01aafb3955216fd0e501 It makes the code more straightforward, in my view, but leads to calls to `safeCtor` even for constructors that don't get chosen in the end.

Reply:
Sure, works for me.
Calls to `createConstructor` are cached, so it should not be a problem.

I've applied your patch **and** removed `inline` from `safeCtor` to reduce the size of the classfile (-25%)